### DISCUSSION TITLE
Cross-platform compilation .Net 5

### DISCUSSION THREAD

**Discussion by neggenbe** [coin-or/Sonnet#24] [Q&A] [answered]
Hi,

I'm trying to use the Sonnet interface in a cross-platform project. Building the project on Windows with .Net 5 works fine but of course, the embedded c++ librairies are not compiled for a Unix environment...

Has this ever been achieved? Of course, there's still the option to pass the model to solve via MPS file, but that would be last the option.

Thanks for the help!
Nik

**Comment by jhmgoossens** [comment]
Hi Nik,
No, I'm not aware of anybody actually using Sonnet cross-platform. 
I've not tried building .NET5 on anything but Windows, let alone the next step of bridging that .NET5 with platform-native C++.  Of course, the embedded COIN-OR C++ libraries of CBC etc. themselves work perfectly fine on non-Windows (since that's how many people use them), but how to then bridge? 
Without solving the native/managed bridge on non-Windows, even merely using Sonnet on non-Windows to create an MPS file won't be easy, since Sonnet uses the native C++ libraries for just about everything, including writing MPS files.
Some googling gave a few links, like [Native interoperability](https://docs.microsoft.com/en-us/dotnet/standard/native-interop/) and [Cross platform C++ in C#](https://stackoverflow.com/questions/65968143/cross-platform-c-in-c-sharp-net-core-5-application). These proposals seem to point to pinvoke / C-interfaces (at first glance). But, since SonnetWrapper uses the object-oriented-ness of the native C++ classes (not the flat pinvoke / C-style) it looks difficult to bridge this. But like I said, never tried. 
Any suggestions or thoughts?

**Comment by jhmgoossens** [comment, marked as answer]
I updated the main README to mention that Sonnet is Windows-only.
The problem is the mixed C++/CLI used in SonnetWrapper. I considered replacing SonnetWrapper by, for example, p/invoke calls to the C API of CoinMP. This would not only be a big design change, but the price would also be the loss of Osi, which currently allows the switching of solvers. Therefore, until further notice, Sonnet remains Windows-only.
